test/e2e/network/no_snat.go

```diff
@@ -52,70 +59,176 @@ var (
-		nodes, err := e2enode.GetReadySchedulableNodes(cs)
+	ginkgo.BeforeEach(func() {
+		var err error
+		cs = f.ClientSet
+		pc = cs.CoreV1().Pods(f.Namespace.Name)
+		nodes = &v1.NodeList{}
+		nodes, err = e2enode.GetBoundedReadySchedulableNodes(cs, maxNodes)
 		framework.ExpectNoError(err)
-		framework.ExpectNotEqual(len(nodes.Items), 0, "no Nodes in the cluster")
+		if len(nodes.Items) < 2 {
+			ginkgo.Skip("At least 2 nodes are required to run the test")
+		}
+
+	})
 
+	ginkgo.It("Should be able to send traffic between Pods without SNAT", func() {
+		framework.Logf("nodes DEBUG %v", nodes)
+		ginkgo.By("creating a test pod on each Node")
+		testPod := createTestPod(testPodPort, false)
+		var wg sync.WaitGroup
 		for _, node := range nodes.Items {
 			// target Pod at Node
+			ginkgo.By("creating pod on node " + node.Name)
 			nodeSelection := e2epod.NodeSelection{Name: node.Name}
 			e2epod.SetNodeSelection(&testPod.Spec, nodeSelection)
-			_, err = pc.Create(context.TODO(), &testPod, metav1.CreateOptions{})
-			framework.ExpectNoError(err)
+			wg.Add(1)
+			go func() {
+				defer ginkgo.GinkgoRecover()
+				defer wg.Done()
+				f.PodClient().CreateSync(testPod)
+			}()
 		}
+		wg.Wait()
 
-		ginkgo.By("waiting for all of the no-snat-test pods to be scheduled and running")
-		err = wait.PollImmediate(10*time.Second, 1*time.Minute, func() (bool, error) {
-			pods, err := pc.List(context.TODO(), metav1.ListOptions{LabelSelector: noSNATTestName})
-			if err != nil {
-				return false, err
-			}
+		ginkgo.By("sending traffic from each pod to the others and checking that SNAT does not occur")
+		pods, err := pc.List(context.TODO(), metav1.ListOptions{LabelSelector: noSNATTestName})
+		framework.ExpectNoError(err)
 
-			// check all pods are running
-			for _, pod := range pods.Items {
-				if pod.Status.Phase != v1.PodRunning {
-					if pod.Status.Phase != v1.PodPending {
-						return false, fmt.Errorf("expected pod to be in phase \"Pending\" or \"Running\"")
-					}
-					return false, nil // pod is still pending
+		// hit the /clientip endpoint on every other Pods to check if source ip is preserved
+		for _, sourcePod := range pods.Items {
+			sourcePod := sourcePod
+			for _, targetPod := range pods.Items {
+				if targetPod.Name == sourcePod.Name {
+					continue
 				}
+				targetAddr := net.JoinHostPort(targetPod.Status.PodIP, testPodPort)
+				ginkgo.By("testing from pod " + sourcePod.Name + " to pod " + targetPod.Name)
+				wg.Add(1)
+				go func() {
+					defer ginkgo.GinkgoRecover()
+					defer wg.Done()
+					sourceIP, execPodIP := execSourceIPTest(sourcePod, targetAddr)
+					ginkgo.By("Verifying the preserved source ip")
+					framework.ExpectEqual(sourceIP, execPodIP)
+				}()
 			}
-			return true, nil // all pods are running
-		})
-		framework.ExpectNoError(err)
+		}
+		wg.Wait()
+	})
+
+	ginkgo.It("Should be able to send traffic between Pods and an agent on that Node without SNAT", func() {
```

Comment:
This isn't testing "an agent on that node" though, it's testing a host-network pod